samples/cellular/nrf_cloud_multi_service/src/provisioning_support.c

```diff
@@ -0,0 +1,184 @@
+
+static bool provisioning_started;
+
+/* Called by the provisioning library to request the LTE modem be taken offline before credential
+ * installation, and then again after credential installation is complete. This is necessary because
+ * credentials cannot be installed to the modem while it is online.
+ *
+ * This limitation only applies to the nrf91 modem; Wi-Fi, for instance, is unaffected. But for
+ * simplicity we use conn_mgr to take take all network interfaces offline and online.
+ */
+static int modem_mode_cb(enum lte_lc_func_mode new_mode, void *user_data)
+{
+	enum lte_lc_func_mode fmode;
+
+	ARG_UNUSED(user_data);
+
+	/* The nrf_provisioning library requires us to return the previous functional mode. */
+	if (lte_lc_func_mode_get(&fmode)) {
+		LOG_ERR("Failed to read modem functional mode");
+		return -EFAULT;
+	}
+
+	if (new_mode == LTE_LC_FUNC_MODE_NORMAL || new_mode == LTE_LC_FUNC_MODE_ACTIVATE_LTE) {
+		LOG_INF("Provisioning library requests normal mode");
+
+		/* Connect all interfaces since we are done provisioning. */
+		conn_mgr_all_if_connect(true);
+
+		/* Wait for network readiness to be re-established before returning. */
+		LOG_DBG("Waiting for network up");
+		k_event_wait(&prov_events, NETWORK_UP, false, K_FOREVER);
+
+		LOG_DBG("Network is up.");
+	} else if (new_mode == LTE_LC_FUNC_MODE_OFFLINE ||
+		   new_mode == LTE_LC_FUNC_MODE_DEACTIVATE_LTE) {
+		LOG_INF("Provisioning library requests offline mode");
+		/* Disconnect all interfaces to allow cert installation. */
+		conn_mgr_all_if_disconnect(true);
+
+		/* Wait for network disconnection before returning. */
+		LOG_DBG("Waiting for network down.");
+		k_event_wait(&prov_events, NETWORK_DOWN, false, K_FOREVER);
+
+		LOG_DBG("Network is down.");
+	}
+
+	return fmode;
+}
+
+/* Delayable work item for marking provisioning as idle. */
+static void mark_provisioning_idle_work_fn(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	LOG_INF("Provisioning is idle.");
+	k_event_post(&prov_events, PROVISIONING_IDLE);
+}
+static K_WORK_DELAYABLE_DEFINE(provisioning_idle_work, mark_provisioning_idle_work_fn);
+
+/* Called by the provisioning library when each provisioning attempt starts, stops, or finishes
+ * with reboot required.
+ */
+static void device_mode_cb(enum nrf_provisioning_event event, void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	switch (event) {
+	case NRF_PROVISIONING_EVENT_START:
+		/* Called when the provisioning library begins checking for provisioning commands.
+		 */
+		LOG_DBG("NRF_PROVISIONING_EVENT_START");
+
+		/* Mark provisioning as active. */
+		LOG_INF("Provisioning is active.");
+		k_event_clear(&prov_events, PROVISIONING_IDLE);
+		break;
+	case NRF_PROVISIONING_EVENT_STOP:
+		/* Called when the current provisioning command check has completed. */
+		LOG_DBG("NRF_PROVISIONING_EVENT_STOP");
+
+		/* Mark provisioning as idle after a small delay.
+		 * This delay is to prevent knocking if the provisioning library performs an
```

Comment:
knocking?

Reply:
Unwanted rapid oscillations of a system, though I cannot find anyone else using the term online, so I'm not sure where I picked it up. It isn't, to my knowledge, my own invention, though.

I can try to rephrase the statement if you think it is unclear. I could replace `knocking` with `false starts`